### PR TITLE
Bring back local garbage bags.

### PR DESCRIPTION
**Pull request by msullivan**
This reverts much of 208bf821e70d432d18b2c1284da4aa2eb7bb46fc although
it also includes a few minor refactors.

### CODE REVIEW

File: src/mem/epoch/garbage.rs [hunk 1]
```diff
@@ -13,7 +13,94 @@ use std::sync::atomic::Ordering::{Relaxed, Release, Acquire};
...
+        self.0 = data;
+    }
+}
+
+// needed because the bags store raw pointers.
+unsafe impl Send for Bag {}
+unsafe impl Sync for Bag {}
+
+/// A thread-local set of garbage bags.
+// FIXME: switch this to use modular arithmetic and accessors instead
```
Comment (aturon) [inline src/mem/epoch/garbage.rs]: Let's drop this FIXME, assuming you agree that the named bags are actually clearer here.

File: src/mem/epoch/garbage.rs [hunk 2]
```diff
@@ -13,7 +13,94 @@ use std::sync::atomic::Ordering::{Relaxed, Release, Acquire};
...
+impl Local {
+    pub fn new() -> Local {
+        Local {
+            old: Bag::new(),
+            cur: Bag::new(),
+            new: Bag::new(),
+        }
+    }
+
+    pub fn reclaim<T>(&mut self, elem: *mut T) {
```
Comment (aturon) [inline src/mem/epoch/garbage.rs]: I'd like to rename this; probably just to `insert`.

**Comment by aturon** [comment]
Looks good, modulo a couple of nits.

**Comment by msullivan** [comment]
Nits addressed

**Comment by msullivan** [comment]
(Am running some more intensive testing right now)

**Comment by msullivan** [comment]
Tests are looking good

**Comment by aturon** [comment]
Woo, merged! And thus (hopefully) ends quite the debugging saga.

**Comment by gnzlbg** [comment]
> Woo, merged! And thus (hopefully) ends quite the debugging saga.

Would something like thread sanitizer have helped here?

**Comment by aturon** [comment]
@gnzlbg It's hard to say. My experience with this kind of thing is that, usually, you just have to stare hard at the code and rethink the algorithm; it tends to be hard to find a meaningful smoking gun.

**Pull request merged**